hal/inc/ble_hal.h

```diff
@@ -15,190 +15,747 @@
+
+/**
+ * Get Characteristic value.
+ *
+ * @param[in]   value_handle    Characteristic value handle.
+ * @param[in]   buf             Pointer to the buffer that to be filled with the Characteristic value.
+ * @param[in]   len             Length of the given buffer. Return the actual length of the value.
+ *
+ * @returns     The length of read data.
+ */
+size_t ble_gatt_server_get_characteristic_value(uint16_t value_handle, uint8_t* buf, size_t len, void* reserved);
+
+/**
+ * Discover all BLE primary services.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ * @param[in]   callback    The callback function to handle the discovered services.
+ * @param[in]   context     The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_client_discover_all_services(uint16_t conn_handle, on_ble_disc_service_cb_t callback, void* context, void* reserved);
+
+/**
+ * Discover BLE service with specific 128-bits service UUID.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ * @param[in]   uuid        Pointer to the service UUID.
+ * @param[in]   callback    The callback function to handle the discovered services.
+ * @param[in]   context     The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_client_discover_service_by_uuid(uint16_t conn_handle, const hal_ble_uuid_t* uuid, on_ble_disc_service_cb_t callback, void* context, void* reserved);
+
+/**
+ * Discover all BLE characteristics within specific handle range.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ * @param[in]   service     The service to be discovered.
+ * @param[in]   callback    The callback function to handle the discovered characteristics.
+ * @param[in]   context     The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_client_discover_characteristics(uint16_t conn_handle, const hal_ble_svc_t* service, on_ble_disc_char_cb_t callback, void* context, void* reserved);
+
+/**
+ * Discover all BLE characteristics within specific handle range.
+ *
+ * @param[in]   conn_handle BLE connection handle.
+ * @param[in]   service     The service to be discovered.
+ * @param[in]   uuid        The specific characteristic UUID.
+ * @param[in]   callback    The callback function to handle the discovered characteristics.
+ * @param[in]   context     The callback function context.
+ *
+ * @returns     0 on success, system_error_t on error.
+ */
+int ble_gatt_client_discover_characteristics_by_uuid(uint16_t conn_handle, const hal_ble_svc_t* service, const hal_ble_uuid_t* uuid, on_ble_disc_char_cb_t callback, void* context, void* reserved);
+
+/**
+ * Check if the BLE discovery procedure is ongoing or not.
+ *
+ * @returns True if BLE is discovering services, characteristics or descriptors, otherwise false.
+ */
+bool ble_gatt_client_is_discovering(void);
```

Comment:
All the other `ble_gatt_client_discover_xxx()` methods have a handle argument. It should probably be added here as well.